### PR TITLE
add CRWriter to add \r needed in raw mode automatically and efficiently

### PR DESCRIPTION
**Pull request by ldemailly**
- [x] move from grol.io/terminal
- [x] add CRWriter to add \r
- [x] add library CI

### CODE REVIEW

File: example/main.go [hunk 1]
```diff
@@ -20,6 +22,26 @@ func Main() int {
...
+	log.SetColorMode()
+	fmt.Fprintf(t.Out, "Terminal is open\nis valid %t\n", isTerm)
+	l, err := t.ReadLine()
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			log.Infof("EOF received, exiting.")
+			return 0
+		}
+		return log.FErrf("Error reading line: %v", err)
+	}
```
Comment (ccoVeille) [inline example/main.go:39]: Easier to read and understand

```suggestion
	if errors.Is(err, io.EOF) {
		log.Infof("EOF received, exiting.")
		return 0
	}
	if err != nil {
		return log.FErrf("Error reading line: %v", err)
	}
```

File: terminal.go [hunk 2]
```diff
@@ -10,13 +11,67 @@ import (
 type Terminal struct {
 	fd       int
 	oldState *term.State
+	term     *term.Terminal
+	Out      io.Writer
+}
+
+// CRWriter is a writer that adds \r before each \n.
+// Needed for raw mode terminals (and I guess also if you want something DOS or http headers like).
```
Comment (ccoVeille) [inline terminal.go:19]:
```suggestion
// Needed for raw mode terminals (and I guess also if you want something DOS or HTTP headers like).
```

File: terminal.go [hunk 3]
```diff
@@ -10,13 +11,67 @@ import (
...
+		io.Reader
+		io.Writer
+	}{os.Stdin, os.Stderr}
+	t := &Terminal{
+		fd:  int(os.Stdin.Fd()),
+		Out: &CRWriter{Out: os.Stdout},
+	}
+	t.term = term.NewTerminal(rw, "")
 	if !t.IsTerminal() {
+		t.Out = os.Stderr // no need to add \r for non raw mode.
```
Comment (ccoVeille) [inline terminal.go:74]:
```suggestion
		t.Out = os.Stderr // no need to add \r for non-raw mode.
```

File: terminal.go [hunk 4]
```diff
@@ -10,13 +11,67 @@ import (
 type Terminal struct {
 	fd       int
 	oldState *term.State
+	term     *term.Terminal
+	Out      io.Writer
+}
```
Comment (ccoVeille) [inline terminal.go:16]:
```suggestion
 type Terminal struct {
	*term.Terminal
 	fd       int
 	oldState *term.State
	Out      io.Writer
}
```
Maybe this, would avoid to define Readline and SetPrompt

File: terminal.go [hunk 5]
```diff
@@ -37,5 +92,14 @@ func (t *Terminal) Close() error {
...
 	return err
 }
+
+func (t *Terminal) ReadLine() (string, error) {
+	return t.term.ReadLine()
+}
+
+func (t *Terminal) SetPrompt(s string) {
+	t.term.SetPrompt(s)
+}
```
Comment (ccoVeille) [inline terminal.go:105]:
```suggestion
```

See https://github.com/fortio/terminal/pull/1#discussion_r1708580332

File: example/main.go [hunk 6]
```diff
@@ -20,6 +22,26 @@ func Main() int {
...
+	log.Infof("Read line got: %q", l)
+	/*
+		if isTerm {
+			cli.UntilInterrupted()
+		}
+	*/
```
Comment (ccoVeille) [inline example/main.go:45]: Wip?

File: go.mod [hunk 7]
```diff
@@ -8,6 +8,8 @@ require (
 	golang.org/x/term v0.23.0
 )
 
+// replace fortio.org/cli => ../cli
```
Comment (ccoVeille) [inline go.mod:11]: Why this?

Comment (ldemailly) [reply go.mod:11]: I was testing https://github.com/fortio/cli/pull/38

File: example/main.go (hunk 6 above)
Comment (ldemailly) [reply example/main.go:45]: yes it's gone now

File: terminal.go (hunk 4 above)
Comment (ldemailly) [reply terminal.go:16]: I'm trying to control the exact exposed surface vs having users know both packages/types

but maybe... tbd

**Pull request merged**